src/main/java/com/google/sps/servlets/SearchServlet.java

```diff
@@ -44,27 +43,34 @@
    */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Initializes a user query object which will seperate the query data for ease of access.
-    UserQuery uQuery = new UserQuery(request.getParameter("query"));
-    
-    // Max size for the query response
-    final int resultsPerRequest = 10;
+    // Initializes a user query object to create the query filters
+    UserQuery uQuery = new UserQuery(
+      request.getParameter("query"),
+      request.getParameter("tags"),
+      request.getParameter("authors")
+    );
     
     // Creates the new Query and adds the filter to be used in Datastore
-    // The filter is currently only set to find an exact match name in Datastore
-    // TODO: Composite Query Filter will be needed to combine and add all filter to query in future
     Query query = new Query("Recipe");
-    Query.Filter publicFilter = new Query.FilterPredicate("published", Query.FilterOperator.EQUAL, true);
-    Query.Filter nameFilter = SearchUtils.getRecipeNameFilter(uQuery.getName());
+    query.setFilter(uQuery.createSearchFilter());
+
+    // We want popular recipes first
+    query.addSort("popularity", Query.SortDirection.DESCENDING);
 
-    query.setFilter(new Query.CompositeFilter(
-      Query.CompositeFilterOperator.AND, Arrays.asList(publicFilter, nameFilter)
-    ));
+    // Changes the size of the response
+    final int resultsToReturn;
+    
+    // If the search request came from the main page, we want the behavior to be different
+    if (request.getHeader("Referer").contains("/pages/MainPage.jsp")) {
+      resultsToReturn = 10;
+    } else {
+      resultsToReturn = 10;
+    }
```

Comment:
Can you add a TODO indicating what might eventually be different for the main page?

Also, you can initialize `resultsToReturn = 10`, and then do the if statement and change it if necessary (you'll have to remove the `final` modifier)

Reply:
Added the TODO, but I left the resultsToReturn the same. I put it as an uninitialized final value, because I thought it would make it clear that the response value should not change in the code/servlet because we want the user to always get the same number of results.

Reply:
I don't fully understand why you're setting it in the if-statements - if you always want it to be the same value anyway, you should define it once as a constant, right?

Reply:
I personally think we should only ever display twelve recipes or potentially multiples of twelve. It's the only way to have the website always have full rows regardless of whether num-columns is 4, 3, 2, or 1.